pkg/virt-api/rest/subresource.go

```diff
@@ -677,6 +677,27 @@ func (app *SubresourceAPIApp) UnfreezeVMIRequestHandler(request *restful.Request
 
 }
 
+func (app *SubresourceAPIApp) SoftRebootVMIRequestHandler(request *restful.Request, response *restful.Response) {
+
+	log.Log.Info("SoftRebootVMIRequestHandler")
+	validate := func(vmi *v1.VirtualMachineInstance) *errors.StatusError {
+		if vmi.Status.Phase != v1.Running {
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("VM is not running"))
+		}
+		condManager := controller.NewVirtualMachineInstanceConditionManager()
+		if condManager.HasConditionWithStatus(vmi, v1.VirtualMachineInstancePaused, v12.ConditionTrue) {
+			return errors.NewConflict(v1.Resource("virtualmachineinstance"), vmi.Name, fmt.Errorf("VMI is paused"))
+		}
+		return nil
```

Comment:
Added here to the validation, the VMI must either have the agent connected condition or ACPI enabled for soft reboot to be possible.

Reply:
updated

Reply:
Note that `vmi.Spec.Domain.Features.ACPI.Enabled` is a `*bool` so it would be safer to test it to be non-nil too.